DIPs/DIP1NNN-MVF.md

```diff
@@ -0,0 +1,159 @@
+
+In the experience report below[4], the effect described above currently backfires on D.  Upon learning D is not `@safe`-by-default, potential D users, rather than questioning their own values as being deficient, perceive D as being deficient.
+
+> At work I've recently changed to a position in which I have more leverage to influence which language certain key parts of our product are written in. Currently that's primarily Java, with a collection of small services written in Go. If I had my way, it'd all be rewritten in D, but I need to convince the other members of my team that it's feasible to use D (as an aside, GC was actually a big plus for them - I work in cybersecurity, so memory safety is a core requirement for most parts of our product. Memory safety *without* a GC would be even better, as there are security concerns around data lifetime with a GC, but baby steps).
+>
+> The first question I was asked was "is it memory safe by default?", to which the answer was painfully "no, only in code annotated with @safe". The inevitable follow-up is "so you have to mark every function with @safe?!"... not a great showing for D in that respect.
+>
+> It's not a deal-breaker, but it already puts D on an uneven footing in regards to Go and Rust, which both claim full memory safety by default.
+
+Transitioning D to a `@safe`-by-default language will demonstrate that D is serious about memory-safety, and is making strides to encourage its proliferation.
+
+### Advantages
+  * D users wanting more robustness and reliability from their programming language will no longer be subject the inconvenience, no matter how small, of opting in to memory safety; D will be memory-safe out of the box.
+  * D will be able to market itself honestly and confidently as a memory-safe language without requiring any addendums equivalent to "some assembly required".
+  * D will be in a better position to compete with other languages that are memory-safe out of the box, and be better prepared for the future as trends turn toward demandng memory safety from software.
+  * Users will have more of a reason to use D over other languages, both memory-safe and memory-unsafe, and more ammunition to advocate for its continued adoption.
+  * The transition to `@safe`-by-default will provide multiple marketing opportunities to get the word out about D and the benefits of its memory safety features.
+
+## Description
+
+This DIP proposes a very simple path forward that will not introduce any breaking changes if users utilize the transition features being proposed.  Within 2 years, starting from the approval of this DIP, D will be `@safe`-by-default.  Existing users, with the flip of a compiler switch, will be able to maintain existing behavior and avoid any breaking changes.  Using the proposed new `version` identifier, `D_SystemByDefault`, users will be able to precisely manage the transition in their code with the full power of D's exceptional meta-programming facilities.
```

Comment:
I think the time should count from the first release implementing the DIP instead. Having a DIP and having a released implementation are two very different things. The users should get 2 full years after the release of the compiler that supports this.

Reply:
> The users should get 2 full years after the release of the compiler that supports this.

I'm sorry, but I don't quite understand.  The would have potentially 100 years after the release of the compiler that supports `@safe`-by-default, the `-revert=safeByDefault` compiler flag, and the `D_SystemByDefault` `version` identifier.  They are all new additions to the compiler and are proposed to remain indefinitely if that's what people want.

Reply:
What do you mean 100 years?

> Within 2 years, starting from the approval of this DIP, D will be `@safe`-by-default.

It says right there in the DIP 2 years after the DIP approval, safe-by-default kicks in.

Reply:
> The users should get 2 full years after the release of the compiler that supports this.

With the statement quoted above, what is it that the users should get 2 full years of?

This DIP proposes that the user will get at least 2 years prior to **Stage 3: Epoch** before they need to do anything. **Stage 1** is just about informing users through documentation and non-verbal communication and is proposed to persist for 1 year.  **Stage 2** introduces compiler flags and a version identifier, but does not make them compulsory, and is proposed to persist that way for 1 year.  Early adopters can get on board in **Stage 2** if they so choose.  At **Stage 3: Epoch** (at least 2 years after the approval of this DIP) users will need to take some kind of action.  Then after **Stage 3: Epoch** there will be 2 years to remove the `-transition=safeByDefault` flag since the compiler will be `@safe`-by-default and that transition flag is of no use.  But after **Stage 3: Epoch**, all other features that were added in **Stage 2** will remain indefinitely unless there is another DIP or grand decision to remove them.

Reply:
> With the statement quoted above, what is it that the users should get 2 full years of?

Stage 2 should persist for 2 years IMHO. This is the transition period where users can (and should) opt-in for safe-by-default without breaking changes. And before you said it was 2 years between the DIP was approved and the introduction of the breaking changes, which if the implementation takes too long could end up with users having very little time to transition.

Now with the minimum time the process should stay in a particular stage, I think the wording is much clearer and safer for users to transition smoothly.

I think it might be even a good idea to introduce some sort of stage 2.5 which is exactly the same as stage 2 but not using `-transition=safeByDefault` or `-revert=safeByDefault` triggers a deprecation warning saying that on version X the compiler will become safe-by-default. This way users can make sure they are prepared for when the switch is flipped. Stage 3 should, of course, remove that deprecation message.

It could be 1 year for stage 2 and 1 year stage 2.5 (or 1.5 years stage 2 and 0.5 years stage 2.5).